### PR TITLE
r/sagemaker_user_profile - new resource

### LINKED ISSUE #14453: Feature: Support for Sagemaker Studio resources
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS has released a fully integrated IDE environment for ML called Sagemaker Studio. API's are now available and support for allocating resources has been added to boto3 for Sagemaker Studio. This request is for the addition of the resources and datasources needed to automate creation of Sagemaker Studio assets.

### New or Affected Resource(s)

* aws_sagemaker_user_profile
* aws_sagemaker_domain
* aws_sagemaker_app
* aws_sagemaker_code_repository

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* https://aws.amazon.com/blogs/aws/amazon-sagemaker-studio-the-first-fully-integrated-development-environment-for-machine-learning/
* https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14453

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_sagemaker_domain - delete implicit EFS file system
resource_aws_sagemaker_user_profile - add new resource


```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerUserProfile_'
--- PASS: TestAccAWSSagemakerUserProfile_basic (362.62s)
--- PASS: TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings (265.57s)
--- PASS: TestAccAWSSagemakerUserProfile_disappears (377.09s)
--- PASS: TestAccAWSSagemakerUserProfile_jupyterServerAppSettings (325.61s)
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettings (274.40s)
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage (335.52s)
--- PASS: TestAccAWSSagemakerUserProfile_tags (357.21s)
```

### CODE REVIEW

File: aws/resource_aws_sagemaker_user_profile.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+					validation.StringLenBetween(1, 63),
+					validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9](-*[a-zA-Z0-9]){0,62}`), "Valid characters are a-z, A-Z, 0-9, and - (hyphen)."),
+				),
+			},
+			"domain_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"single_sign_on_user_indentifier": {
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:48]:
```suggestion
			"single_sign_on_user_identifier": {
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+	input := &sagemaker.CreateUserProfileInput{
+		UserProfileName: aws.String(d.Get("user_profile_name").(string)),
+		DomainId:        aws.String(d.Get("domain_id").(string)),
+	}
+
+	if v, ok := d.GetOk("user_settings"); ok {
+		input.UserSettings = expandSagemakerDomainDefaultUserSettings(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("single_sign_on_user_indentifier"); ok {
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:230]:
```suggestion
	if v, ok := d.GetOk("single_sign_on_user_identifier"); ok {
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+			log.Printf("[WARN] Unable to find SageMaker UserProfile (%s), removing from state", d.Id())
+			return nil
+		}
+		return fmt.Errorf("error reading SageMaker UserProfile (%s): %w", d.Id(), err)
+	}
+
+	arn := aws.StringValue(UserProfile.UserProfileArn)
+	d.Set("user_profile_name", UserProfile.UserProfileName)
+	d.Set("domain_id", UserProfile.DomainId)
+	d.Set("single_sign_on_user_indentifier", UserProfile.SingleSignOnUserIdentifier)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:285]:
```suggestion
	d.Set("single_sign_on_user_identifier", UserProfile.SingleSignOnUserIdentifier)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	domainID, userProfileName, err := decodeSagemakerUserProfileName(d.Id())
+	if err != nil {
+		return err
+	}
+
+	UserProfile, err := finder.UserProfileByName(conn, domainID, userProfileName)
+	if err != nil {
+		if isAWSErr(err, sagemaker.ErrCodeResourceNotFound, "") {
+			d.SetId("")
+			log.Printf("[WARN] Unable to find SageMaker UserProfile (%s), removing from state", d.Id())
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:276]:
```suggestion
			log.Printf("[WARN] Unable to find SageMaker User Profile (%s), removing from state", d.Id())
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+			return nil
+		}
+		return fmt.Errorf("error reading SageMaker UserProfile (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:279]:
```suggestion
		return fmt.Errorf("error reading SageMaker User Profile (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+	arn := aws.StringValue(UserProfile.UserProfileArn)
+	d.Set("user_profile_name", UserProfile.UserProfileName)
+	d.Set("domain_id", UserProfile.DomainId)
+	d.Set("single_sign_on_user_indentifier", UserProfile.SingleSignOnUserIdentifier)
+	d.Set("single_sign_on_user_value", UserProfile.SingleSignOnUserValue)
+	d.Set("arn", arn)
+	d.Set("home_efs_file_system_uid", UserProfile.HomeEfsFileSystemUid)
+
+	if err := d.Set("user_settings", flattenSagemakerDomainDefaultUserSettings(UserProfile.UserSettings)); err != nil {
+		return fmt.Errorf("error setting user_settings for SageMaker UserProfile (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:291]:
```suggestion
		return fmt.Errorf("error setting user_settings for SageMaker User Profile (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+	tags, err := keyvaluetags.SagemakerListTags(conn, arn)
+
+	if err != nil {
+		return fmt.Errorf("error listing tags for SageMaker UserProfile (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:297]:
```suggestion
		return fmt.Errorf("error listing tags for SageMaker User Profile (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+		input := &sagemaker.UpdateUserProfileInput{
+			UserProfileName: aws.String(d.Get("user_profile_name").(string)),
+			DomainId:        aws.String(d.Get("domain_id").(string)),
+			UserSettings:    expandSagemakerDomainDefaultUserSettings(d.Get("user_settings").([]interface{})),
+		}
+
+		log.Printf("[DEBUG] SageMaker User Profile update config: %#v", *input)
+		_, err := conn.UpdateUserProfile(input)
+		if err != nil {
+			return fmt.Errorf("error updating SageMaker UserProfile: %w", err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:320]:
```suggestion
			return fmt.Errorf("error updating SageMaker User Profile: %w", err)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+	domainID := d.Get("domain_id").(string)
+
+	input := &sagemaker.DeleteUserProfileInput{
+		UserProfileName: aws.String(userProfileName),
+		DomainId:        aws.String(domainID),
+	}
+
+	if _, err := conn.DeleteUserProfile(input); err != nil {
+		if !isAWSErr(err, sagemaker.ErrCodeResourceNotFound, "") {
+			return fmt.Errorf("error deleting SageMaker UserProfile (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:348]:
```suggestion
			return fmt.Errorf("error deleting SageMaker User Profile (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+		}
+	}
+
+	if _, err := waiter.UserProfileDeleted(conn, domainID, userProfileName); err != nil {
+		if !isAWSErr(err, sagemaker.ErrCodeResourceNotFound, "") {
+			return fmt.Errorf("error waiting for SageMaker UserProfile (%s) to delete: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:354]:
```suggestion
			return fmt.Errorf("error waiting for SageMaker User Profile (%s) to delete: %w", d.Id(), err)
```

File: website/docs/r/sagemaker_user_profile.html.markdown [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `user_profile_name` - (Required) The name for the User Profile.
+* `domain_id` - (Required) The ID of the associated Domain.
+* `single_sign_on_user_indentifier` - (Optional) A specifier for the type of value specified in `single_sign_on_user_value`. Currently, the only supported value is `UserName`. If the Domain's AuthMode is SSO, this field is required. If the Domain's AuthMode is not SSO, this field cannot be specified.
```
Comment (ewbankkit) [inline website/docs/r/sagemaker_user_profile.html.markdown:30]:
```suggestion
* `single_sign_on_user_identifier` - (Optional) A specifier for the type of value specified in `single_sign_on_user_value`. Currently, the only supported value is `UserName`. If the Domain's AuthMode is SSO, this field is required. If the Domain's AuthMode is not SSO, this field cannot be specified.
```

File: website/docs/r/sagemaker_user_profile.html.markdown [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+* `single_sign_on_user_value` - (Required) The username of the associated AWS Single Sign-On User for this UserProfile. If the Domain's AuthMode is SSO, this field is required, and must match a valid username of a user in your directory. If the Domain's AuthMode is not SSO, this field cannot be specified.
```
Comment (ewbankkit) [inline website/docs/r/sagemaker_user_profile.html.markdown:31]:
```suggestion
* `single_sign_on_user_value` - (Required) The username of the associated AWS Single Sign-On User for this User Profile. If the Domain's AuthMode is SSO, this field is required, and must match a valid username of a user in your directory. If the Domain's AuthMode is not SSO, this field cannot be specified.
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+				),
+			},
+			"domain_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"single_sign_on_user_indentifier": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:50]: Add `ForceNew` as cannot be updated.

```suggestion
				Type:     schema.TypeString,
				Optional: true,
				ForceNew: true,
```

File: aws/resource_aws_sagemaker_user_profile.go [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+			},
+			"single_sign_on_user_value": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_user_profile.go:54]: Add `ForceNew` as cannot be updated.

```suggestion
				Type:     schema.TypeString,
				Optional: true,
				ForceNew: true,
```

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSagemakerUserProfile_' ACCTEST_PARALLELISM=1
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAWSSagemakerUserProfile_ -timeout 120m
=== RUN   TestAccAWSSagemakerUserProfile_basic
=== PAUSE TestAccAWSSagemakerUserProfile_basic
=== RUN   TestAccAWSSagemakerUserProfile_tags
=== PAUSE TestAccAWSSagemakerUserProfile_tags
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== RUN   TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_disappears
=== PAUSE TestAccAWSSagemakerUserProfile_disappears
=== CONT  TestAccAWSSagemakerUserProfile_basic
--- PASS: TestAccAWSSagemakerUserProfile_basic (281.59s)
=== CONT  TestAccAWSSagemakerUserProfile_disappears
--- PASS: TestAccAWSSagemakerUserProfile_disappears (250.77s)
=== CONT  TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_jupyterServerAppSettings (283.07s)
=== CONT  TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings (283.22s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage (194.54s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettings (210.70s)
=== CONT  TestAccAWSSagemakerUserProfile_tags
--- PASS: TestAccAWSSagemakerUserProfile_tags (252.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1756.184s
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSagemakerUserProfile_' ACCTEST_PARALLELISM=1
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAWSSagemakerUserProfile_ -timeout 120m
=== RUN   TestAccAWSSagemakerUserProfile_basic
=== PAUSE TestAccAWSSagemakerUserProfile_basic
=== RUN   TestAccAWSSagemakerUserProfile_tags
=== PAUSE TestAccAWSSagemakerUserProfile_tags
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== RUN   TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_disappears
=== PAUSE TestAccAWSSagemakerUserProfile_disappears
=== CONT  TestAccAWSSagemakerUserProfile_basic
--- PASS: TestAccAWSSagemakerUserProfile_basic (303.38s)
=== CONT  TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings (204.26s)
=== CONT  TestAccAWSSagemakerUserProfile_disappears
--- PASS: TestAccAWSSagemakerUserProfile_disappears (380.98s)
=== CONT  TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_jupyterServerAppSettings (199.93s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettings (210.40s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage (201.35s)
=== CONT  TestAccAWSSagemakerUserProfile_tags
--- PASS: TestAccAWSSagemakerUserProfile_tags (334.45s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1834.835s
```

**Comment by DrFaust92** [comment]
Rebased onto master

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSSagemakerUserProfile_' ACCTEST_PARALLELISM=1
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAWSSagemakerUserProfile_ -timeout 120m
=== RUN   TestAccAWSSagemakerUserProfile_basic
=== PAUSE TestAccAWSSagemakerUserProfile_basic
=== RUN   TestAccAWSSagemakerUserProfile_tags
=== PAUSE TestAccAWSSagemakerUserProfile_tags
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== PAUSE TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
=== RUN   TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== PAUSE TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
=== RUN   TestAccAWSSagemakerUserProfile_disappears
=== PAUSE TestAccAWSSagemakerUserProfile_disappears
=== CONT  TestAccAWSSagemakerUserProfile_basic
--- PASS: TestAccAWSSagemakerUserProfile_basic (393.87s)
=== CONT  TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_kernelGatewayAppSettings (313.66s)
=== CONT  TestAccAWSSagemakerUserProfile_disappears
--- PASS: TestAccAWSSagemakerUserProfile_disappears (253.13s)
=== CONT  TestAccAWSSagemakerUserProfile_jupyterServerAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_jupyterServerAppSettings (350.13s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettings
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettings (187.27s)
=== CONT  TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage
--- PASS: TestAccAWSSagemakerUserProfile_tensorboardAppSettingsWithImage (208.58s)
=== CONT  TestAccAWSSagemakerUserProfile_tags
--- PASS: TestAccAWSSagemakerUserProfile_tags (228.44s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1936.854s
```

As user profile is part of SageMaker studio, it will not work in GovCloud. A tracking issue will be linked to this PR which will cover updating the tests to SKIP correctly in that partiion.

**Pull request merged**